src/triage/tracking.py

```diff
@@ -0,0 +1,282 @@
+            'http://169.254.169.254/latest/meta-data/instance-type',
+            timeout=1
+        ).text
+    except requests.exceptions.ConnectionError:
+        logging.info(
+            "Unable to retrieve metadata about ec2 instance, will not set ec2 instance type"
+        )
+        ec2_instance_type = None
+    return ec2_instance_type
+
+
+def infer_log_location():
+    """Attempt to infer the location of the log file of the root logger
+
+    Returns: Either the baseFilename of the first FileHandler on the root logger, or None
+    """
+    root_logger_handlers = [
+        handler
+        for handler in logging.getLoggerClass().root.handlers
+        if isinstance(handler, logging.FileHandler)
+    ]
+    if root_logger_handlers:
+        log_location = root_logger_handlers[0].baseFilename
+    else:
+        logging.info("No FileHandler found in root logger, cannot record logging filename")
+        log_location = None
+    return log_location
+
+
+def initialize_tracking_and_get_run_id(
+    experiment_hash,
+    experiment_class_name,
+    experiment_kwargs,
+    db_engine
+):
+    """Create a row in the ExperimentRun table with some initial info and return the created run_id
+
+    Args:
+        experiment_hash (str) An experiment hash that exists in the experiments table
+        experiment_class_name (str) The name of the experiment subclass used
+        experiment_kwargs (dict) Any runtime Experiment keyword arguments that should be saved
+        db_engine (sqlalchemy.engine)
+    """
+    # Any experiment kwargs that are types (e.g. MatrixStorageClass) can't
+    # be serialized, so just use the class name if so
+    cleaned_experiment_kwargs = {
+        k: (v.__name__ if isinstance(v, type) else v)
+        for k, v in experiment_kwargs.items()
+    }
+    run = ExperimentRun(
+        start_time=datetime.datetime.now(),
+        git_hash=infer_git_hash(),
+        triage_version=infer_triage_version(),
+        experiment_hash=experiment_hash,
+        last_updated_time=datetime.datetime.now(),
+        current_status=ExperimentRunStatus.started,
+        installed_libraries=infer_installed_libraries(),
+        platform=platform.platform(),
+        os_user=getpass.getuser(),
+        working_directory=os.getcwd(),
+        ec2_instance_type=infer_ec2_instance_type(),
+        log_location=infer_log_location(),
+        experiment_class_name=experiment_class_name,
+        experiment_kwargs=cleaned_experiment_kwargs,
+        matrices_made=0,
+        matrices_errored=0,
+        matrices_skipped=0,
+        models_made=0,
+        models_errored=0,
+        models_skipped=0,
+    )
+    run_id = None
+    with scoped_session(db_engine) as session:
+        session.add(run)
+        session.commit()
+        run_id = run.run_id
+    if not run_id:
+        raise ValueError("Failed to retrieve run_id from saved row")
+    return run_id
+
+
+@contextmanager
+def get_run_obj_for_update(db_engine, run_id):
```

Comment:
I imagine this could be a composition:

```
@contextmanager
def get_run_for_update(session_manager, run_id):
    with triage.util.db.get_for_update(session_manager, ExperimentRun, run_id) as run:
        yield run
        run.last_updated_time = datetime.datetime.now()
```

That said, can't you just set `(onupdate=datetime.datetime.now)` for this column?

[[docs]](https://docs.sqlalchemy.org/en/latest/core/defaults.html)